### DISCUSSION TITLE
Enable root ssh access to LXC

### DISCUSSION THREAD

**Discussion by vdias** [tteck/Proxmox#385] [Ideas]
Enable root ssh access to LXC.

**Comment by tteck** [comment]
1) If you use auto login, set a root password `sudo passwd root`
2) `nano /etc/ssh/sshd_config` replace `PermitRootLogin prohibit-password` with `PermitRootLogin yes`
3) `systemctl restart sshd`

**Reply by ruhanirabin** [reply]
If you want to do it in a single line command **using your _password as a parameter_**, this script line may reduce using your keyboard and manual input. please note this is not intended to use outside internal network environments

You just modify the part ➡️ **`Your@StrongPassword`**

```bash
sudo sh -c 'echo "root:$1" | chpasswd && sed -i "/^#*PermitRootLogin/c\PermitRootLogin yes" /etc/ssh/sshd_config; grep -q "^PermitRootLogin" /etc/ssh/sshd_config || echo "PermitRootLogin yes" >> /etc/ssh/sshd_config && systemctl restart sshd' -- "Your@StrongPassword"
```

This will check if there 
- is a commented line with `PermitRootLogin`
- if `yes` is not set as the part of `PermitRootLogin`
- if the line `Permitrootlogin `is missing from the file

_This will enable Root login (gracefully) and reload SSHD accordingly so your next login will require password to login._

this was tested on Debian 12 LXCs. you can modify it as per your needs. I hope it helps someone. Thanks @tteck

**Comment by hellofaduck** [comment]
Maybe you can add this option to container creation gui, because it's so frustrating to login through proxmox gui to every created container, edit sshd config, restart it etc. It's not a big deal, but we here because _automation_, isn't it? :)

**Reply by tteck** [reply]
I think the decision to `PermitRootLogin yes` should be made by the end user.

**Reply by hellofaduck** [reply]
I agree, its very sensitive option, because of that i suggest you to add an option at container creation stage to choose that with gui

**Reply by tteck** [reply]
Oh, yes you did! lol

**Reply by tteck** [reply]
Testing in the Debian LXC

**Reply by tteck** [reply]
Now available for all LXC's

**Reply by hellofaduck** [reply]
So awesome! Thank you for cooperation! :)

**Reply by danielo515** [reply]
How? I never got the question

**Reply by tteck** [reply]
Advanced settings

**Reply by bilogic** [reply]
can SSH_ROOT be modified from the CLI? then there is no need to answer every question in GUI

**Reply by tteck** [reply]
If only you had read the thread...
https://github.com/tteck/Proxmox/discussions/385#discussioncomment-3283416

**Reply by bilogic** [reply]
I did read. I meant during the installation, e.g.

```
SSH_ROOT=yes bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/mariadb.sh)"
```

Instead of going into `Advanced settings`, because some of the questions there are not easy to answer for first timers and the defaults are not pre-selected, e.g. selecting a wrong OS can cause the LXC not to work

**Reply by tteck** [reply]
nope. It tells you what the default OS is.

**Reply by tteck** [reply]
![Screenshot 2024-09-01 7 25 18 AM](https://github.com/user-attachments/assets/bfc2414a-9ede-4dbe-9244-2cd37eddcfa4)

**Reply by bilogic** [reply]
OS is just an example, nevermind thanks

**Reply by bilogic** [reply]
This is a digression from this discussion: but if only 1 OS is allowed, why still ask the user?

**Reply by tteck** [reply]
You can try any distribution you want, but I only provide script support if the default distribution is used.

e.g. Ubuntu 24.04 is used
```
    __  ___           _       ____  ____ 
   /  |/  /___ ______(_)___ _/ __ \/ __ )
  / /|_/ / __  / ___/ / __  / / / / __  |
 / /  / / /_/ / /  / / /_/ / /_/ / /_/ / 
/_/  /_/\__,_/_/  /_/\__,_/_____/_____/  
 
Using Advanced Settings
Using Distribution: ubuntu
Using ubuntu Version: 24.04
Using Container Type: 1
Using Root Password: Automatic Login
Container ID: 111
Using Hostname: mariadb
Using Disk Size: 4
Allocated Cores: 1
Allocated RAM: 1024
Using Bridge: vmbr0
Using IP Address: dhcp
Using Gateway IP Address: Default
Using APT-Cacher IP Address: Default
Disable IPv6: no
Using Interface MTU Size: Default
Using DNS Search Domain: Host
Using DNS Server IP Address: Host
Using MAC Address: AE:1A:60:07:B7:59
Using Vlan: Default
Enable Root SSH Access: no
Enable Verbose Mode: no
Creating a MariaDB LXC using the above advanced settings
 ✓ Using local for Template Storage.
 ✓ Using thin-pool for Container Storage.
 ✓ Updated LXC Template List
 ✓ LXC Container 111 was successfully created.
 ✓ Started LXC Container
 ✓ Set up Container OS
 ✓ Network Connected: 192.168.86.250 
 ✓ IPv4 Internet Connected
 ✗ IPv6 Internet Not Connected
 ✓ DNS Resolved github.com to 140.82.112.3
 ✓ Updated Container OS
 ✓ Installed Dependencies
 ✓ Installed MariaDB
 ✓ Customized Container
 ✓ Cleaned
 ✓ Completed Successfully!
```
works fine.

**Reply by bilogic** [reply]
@tteck i think we are completely digressing, my question/suggestion was: can we customize the installer without answering questions in the GUI, because I do want to stick to your tested defaults, but just add on SSH root access

**Reply by tteck** [reply]
and I answered, nope

**Reply by tteck** [reply]
> This is a digression from this discussion: but if only 1 OS is allowed, why still ask the user?

And, I was answering your question.

**Reply by bilogic** [reply]
> > This is a digression from this discussion: but if only 1 OS is allowed, why still ask the user?
> 
> And, I was answering your question.

There was 1 script (not MariaDB) and I cannot remember which one, where only Debian works.

As a more concrete example, after selecting Debian, I'm asked which version 11/12... which I have no idea how to answer.

**Reply by tteck** [reply]
There's more than one that way. 🤷🏻‍♂️  It is called "Advanced". If using advanced settings feels uncomfortable, and you only need SSH, it's best to stick with the method described above. I'm not sure where this conversation is heading.

**Reply by bilogic** [reply]
Well, you have answered no to my main question. Everything else was just to support why I raised the main question, I was not seeking answers to those.

**Reply by tteck** [reply]
> As a more concrete example, after selecting Debian, I'm asked which version 11/12... which I have no idea how to answer.

I've already answered this! Each script when using advanced settings will state the default. https://github.com/tteck/Proxmox/discussions/385#discussioncomment-10511726

**Reply by ruhanirabin** [reply]
Hats off to @tteck . I always use debian 12 for the LXC and all of them (20x) working and stable. I am planning to use ansible but not sure ansible is compatible with LXC (it should be, by theory) - that is why I am here and he already answered the question of SSH login enable option manually which i will go through.
